packages/next-server/server/api-utils.ts

```diff
@@ -0,0 +1,169 @@
+import { IncomingMessage } from 'http'
+import { NextApiResponse, NextApiRequest } from '../lib/utils'
+import { Stream } from 'stream'
+import getRawBody from 'raw-body'
+import { URL } from 'url'
+
+/**
+ * Parse incoming message like `json` or `urlencoded`
+ * @param req
+ */
+export async function parseBody(req: NextApiRequest) {
+  const type = req.headers['content-type'] || 'text/plain'
+  const encoding = getCharset(type)
+
+  const buffer = await getRawBody(req, { encoding })
+
+  const body = buffer.toString()
+
+  if (type.startsWith('application/json')) {
+    return parseJson(body)
+  } else if (type.startsWith('application/x-www-form-urlencoded')) {
+    const qs = require('querystring')
+    return qs.decode(body)
+  } else {
+    return body
+  }
+}
+
+/**
+ * Parse `JSON` and handles invalid `JSON` strings
+ * @param str `JSON` string
+ */
+function parseJson(str: string) {
+  try {
+    return JSON.parse(str)
+  } catch (e) {
+    throw new ApiError(400, 'Invalid JSON')
+  }
+}
+
+/**
+ * Parsing query arguments from request `url` string
+ * @param url of request
+ * @returns Object with key name of query argument and its value
+ */
+export function parseQuery({ url, headers }: IncomingMessage) {
+  if (url) {
+      const params = new URL(`${headers.host}${url}`).searchParams
```

Comment:
Why is the host header being used. `req.url` should be enough.

Reply:
`URL` needs to have full url.

Reply:
But why are you using `URL` 🤔

Reply:
I wanted to use `url` but it's going to be deprecated https://nodejs.org/docs/latest/api/url.html#url_legacy_url_api

Reply:
Also, by doing it this way:
```js
const params = new URL(url, 'http://x').searchParams;
```
`url` can now be both absolute and relative